test: Introduce `SUPPRESS_ABORT_MESSAGE` environment variable

hebasto opened the pull request:
This PR adds a way to suppress the abort message box when running `test_bitcoin.exe` and `fuzz.exe` built with the debug runtime library on Windows.

Otherwise, a failing `assert()` triggers the `abort` routine, which displays a message box and causes a timeout in CI.

Here are CI jobs for the "Debug" configuration:
 - on the master branch: https://github.com/maflcko/bitcoin-core-with-ci/actions/runs/14759346479/job/41435857332
 - with this PR: https://github.com/hebasto/bitcoin-core-nightly/actions/runs/15038743417/job/42265371841

Addresses this [comment](https://github.com/bitcoin/bitcoin/issues/32341#issuecomment-2842716175).

According to Microsoft's [docs](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/abort), this behavior is MSCV-specific.

shahsb commented on src/test/fuzz/fuzz.cpp:
```diff
@@ -245,6 +245,10 @@ int main(int argc, char** argv)
         test_one_input(buffer);
         return 0;
     }
+#ifdef WIN32
```
```suggestion
#ifdef _WIN32
```

`_WIN32` is a compiler flag and could be used for both windows platform i.e windows_x86 (32-bit) as well as windows_x64 (64-bit)

hebasto replied:
The current code is consistent with our entire code base.

shahsb commented:
Concept ACK

maflcko reviewed (commented):
No objection, but assert/Assert seems to be used widely in the codebase, so shouldn't this be done for all test binaries, or none? Otherwise the same assert could lead to inconsistent behavior, depending on which test binary ran into it.

hebasto commented:
@maflcko 

> No objection, but assert/Assert seems to be used widely in the codebase, so shouldn't this be done for all test binaries, or none? Otherwise the same assert could lead to inconsistent behavior, depending on which test binary ran into it.

Thanks! Reworked.

hebasto commented:
Not related to this PR, but the [error in the `wallet_fees` target](https://github.com/hebasto/bitcoin-core-nightly/actions/runs/14801894543/job/41562322239#step:8:484):
```
wallet_fees: succeeded against 23 files in 0s.
Run wallet_fees with args ['D:\\a\\bitcoin-core-nightly\\bitcoin-core-nightly\\build\\bin\\Debug\\fuzz.exe', WindowsPath('D:/a/_temp/qa-assets/fuzz_corpora/wallet_fees')]
\u26a0\ufe0f Failure generated from target with exit code 3221225477: ['D:\\a\\bitcoin-core-nightly\\bitcoin-core-nightly\\build\\bin\\Debug\\fuzz.exe', WindowsPath('D:/a/_temp/qa-assets/fuzz_corpora/wallet_fees')]
```

can be fixed with the following patch:
```diff
--- a/src/wallet/test/fuzz/fees.cpp
+++ b/src/wallet/test/fuzz/fees.cpp
@@ -15,22 +15,22 @@
 namespace wallet {
 namespace {
 const TestingSetup* g_setup;
-static std::unique_ptr<CWallet> g_wallet_ptr;
 
 void initialize_setup()
 {
     static const auto testing_setup = MakeNoLogFileContext<const TestingSetup>();
     g_setup = testing_setup.get();
-    const auto& node{g_setup->m_node};
-    g_wallet_ptr = std::make_unique<CWallet>(node.chain.get(), "", CreateMockableWalletDatabase());
 }
 
 FUZZ_TARGET(wallet_fees, .init = initialize_setup)
 {
+    SeedRandomStateForTest(SeedRand::ZEROS);
     FuzzedDataProvider fuzzed_data_provider{buffer.data(), buffer.size()};
+    SetMockTime(ConsumeTime(fuzzed_data_provider));
     const auto& node{g_setup->m_node};
     Chainstate* chainstate = &node.chainman->ActiveChainstate();
-    CWallet& wallet = *g_wallet_ptr;
+    std::unique_ptr<CWallet> wallet_ptr{std::make_unique<CWallet>(node.chain.get(), "", CreateMockableWalletDatabase())};
+    CWallet& wallet{*wallet_ptr};
     {
         LOCK(wallet.cs_wallet);
         wallet.SetLastBlockProcessed(chainstate->m_chain.Height(), chainstate->m_chain.Tip()->GetBlockHash());
```

cc @dergoegge @marcofleon

fanquake reviewed (commented):
Not sure about compiler specific changes, that rely on the presence of an undocumented environment variable (which we don't explicitly set as far as I can tell?), to make tests work properly.

hebasto commented on src/test/util/setup_common.cpp:
```diff
@@ -73,6 +74,18 @@ using node::LoadChainstate;
 using node::RegenerateCommitments;
 using node::VerifyLoadedChainstate;
 
+#if defined(_MSC_VER)
```
This behaviour is MSVC-specific (I've added a note to the PR description). And I cannot reproduce it when running cross-compiled binaries.

fanquake replied:
I think the commit message should still be fixed to say `When compiling with MSVC`, rather than `On Windows`?

> (I've added a note to the PR description)

Is it a bug in mingw-w64 that it doesn't match the MSVC behaviour? It would be good to an explanation of the difference in the PR description/commit message, as it isn't really clear from going to the link you've provided.

hebasto replied:
> Is it a bug in mingw-w64 that it doesn't match the MSVC behaviour?

I don't think so, given the different runtime libraries used. I must admit I haven't tested cross-compiled binaries linked to UCRT.

fanquake commented:
> It is documented, and it is pre-defined by CI itself:

It's not in our docs, or in our CI code, which means our CI would "work", because a third party is putting something into the environment, and the fact that this is even happening, is only discoverable if you happen to read a `.cpp` file.

Would someone running these binaries locally also need/want to set this env var to get the same behaviour? If so, how would they figure that out?

hebasto commented:
> > It is documented, and it is pre-defined by CI itself:
> 
> It's not in our docs, or in our CI code, which means our CI would "work", because a third party is putting something into the environment, and the fact that this is even happening, is only discoverable if you happen to read a `.cpp` file.

What are you suggesting should be changed to improve or avoid this situation?

> Would someone running these binaries locally also need/want to set this env var to get the same behaviour? If so, how would they figure that out?

Only CI jobs defined by scripts in the `ci/test/` directory are intended to be runnable locally. This does not apply to the "raw" GitHub Actions workflows, where this code takes effect.

fanquake commented:
> This does not apply to the "raw" GitHub Actions workflows, where this code takes effect.

Can't someone just compile and run the same binaries locally?

hebasto commented:
> > This does not apply to the "raw" GitHub Actions workflows, where this code takes effect.
> 
> Can't someone just compile and run the same binaries locally?

Certainly, they can. But displaying a message box isn't an issue when running binaries locally, so there's no need to alter the environment in that case.

hebasto commented:
As an alternative, we could explicitly set our own `SUPPRESS_ABORT_MESSAGE_BOX` environment variable in the workflow, and gate the code with this variable instead, removing the `#if defined(_MSC_VER)` condition.

hebasto committed:
test: Introduce `SUPPRESS_ABORT_MESSAGE` environment variable

This change adds a way to suppress the abort message box when running
test or fuzz executables built with the debug runtime library on
Windows. Without this, the message box may cause timeouts in CI.

hebasto commented:
> > It is documented, and it is pre-defined by CI itself:
> 
> It's not in our docs, or in our CI code, which means our CI would "work", because a third party is putting something into the environment, and the fact that this is even happening, is only discoverable if you happen to read a `.cpp` file.
> 
> Would someone running these binaries locally also need/want to set this env var to get the same behaviour? If so, how would they figure that out?

Reworked to avoid relying on third-party set environment variables.

hebasto commented:
This issue seems related: https://developercommunity.visualstudio.com/t/Debug-CRT-shows-abort-message-box-instea/10904782.

fanquake commented:
> This PR adds a way to suppress the abort message box when running test_bitcoin.exe and fuzz.exe built with the debug runtime library on Windows.

Given this isn't being used anywhere in our CI, it's no-longer clear why this is needed (we don't generally add unused code to the project). This also still isn't documented anywhere, so it's not clear how anyone who would need to use this, would know that it exists.

hodlinator commented:
Concept ACK as long as we run binaries built with the proprietary Microsoft runtime on CI.

GUI assert MessageBoxes on Windows are the bane of CI immemorial.

I might be less positive *if* UCRT turns out to not exhibit this behavior (https://github.com/bitcoin/bitcoin/pull/32409#discussion_r2081956577) and we drop MSVCRT. Or if we switch to clang-cl and no longer have the issue.

fanquake commented:
> Not sure what you are getting at

The ENV var isn't added to any CI here. So this is currently just unused code.

hodlinator commented:
> The ENV var isn't added to any CI here. So this is currently just unused code.

Alright, I agree that *only* adding this code will have no affect.

Am I right to understand that the reason for us not setting `SUPPRESS_ABORT_MESSAGE` in the Windows native jobs in this PR is that we're never using the Debug runtime in our in-tree CI configuration?
That would mean this is only intended to be used for out-of-tree nightly/custom configurations.

I don't see much reason for guarding it behind an env-variable? Guarding behind `_MSC_VER` and possibly also `_DEBUG` seems sufficient to me (Edit: since Windows debug builds tested by non-devs are probably Guix generated non-MSVC anyway, so there's no loss in missing GUI MessageBoxes). If you feel uncomfortable to rely on custom out-of-tree CI and rare Windows devs to build it, I kind of see your point.

sedited reviewed (commented):
I'm also not sure about adding this without anything actually using it. hodlinator's suggestion of just dropping the env variable fine to me.

hebasto commented:
> > It is already implemented for `bitcoind.exe` here:
> 
> Ah interesting. Would it be possible to drop that with clang-cl.exe ?

Unfortunately not. This behavior is dictated by the Microsoft C Runtime rather than the compiler itself.

hebasto commented on src/test/util/setup_common.cpp:
```diff
@@ -74,6 +73,23 @@ using node::LoadChainstate;
+// box may cause timeouts.
+// To suppress it, define the `SUPPRESS_ABORT_MESSAGE` environment variable.
+// See: https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/abort
+#include <stdlib.h>
```
According to https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/set-abort-behavior, `<stdlib.h>` provides `_set_abort_behavior`.

hebasto commented:
> I'm also not sure about adding this without anything actually using it.

https://github.com/hebasto/bitcoin-core-nightly/actions/workflows/windows-msvc-x86_64.yml

If we conclude that this doesn't count as an actual use case, then we should probably remove this code as well:https://github.com/bitcoin/bitcoin/blob/5336bcd5784925cd722cff4b9c28e58c64c0b708/src/init.cpp#L863-L864

fanquake commented:
> https://github.com/hebasto/bitcoin-core-nightly/actions/workflows/windows-msvc-x86_64.yml

I can't tell from this link if you mean you are using this already, or would use this later, or something else, like just setting `SUPPRESS_ABORT_MESSAGE` in the nightly env?

Also is this obsoleted with #33593?
